### PR TITLE
Bugfix for issue #174

### PR DESCRIPTION
**Pull request by k-tokarev**
In order to support my issue (https://github.com/reactor/reactor-pool/issues/174#issue-1875620073) I made this pull request with test indicating the problem and my "naive" fix for it. Project builds successfully with this fix, so I hope it don't brake anything.

### CODE REVIEW

**Comment by k-tokarev** [comment]
Sorry for imports rearranged in this pull request. My IntelliJ IDEA did it automatically and I don't know how to disable this for one single project.

File: reactor-pool/src/main/java/reactor/pool/SimpleDequePool.java [hunk 1]
```diff
@@ -29,22 +40,6 @@
...
-import reactor.core.Exceptions;
-import reactor.core.Scannable;
-import reactor.core.publisher.Flux;
-import reactor.core.publisher.Mono;
-import reactor.core.publisher.Operators;
-import reactor.core.scheduler.Scheduler;
-import reactor.core.scheduler.Schedulers;
-import reactor.util.Loggers;
-import reactor.util.annotation.Nullable;
-
```
Comment (pderop) [inline reactor-pool/src/main/java/reactor/pool/SimpleDequePool.java:47]: can you please revert the imports in order to reduce the impact of the patch ?

File: reactor-pool/src/main/java/reactor/pool/SimpleDequePool.java [hunk 2]
```diff
@@ -451,7 +446,7 @@ else if (sig.isOnError()) {
 							final int mergeConcurrency = Math.min(poolConfig.allocationStrategy().warmupParallelism(), toWarmup + 1);
 							Flux.range(1, toWarmup)
 									.map(i -> warmupMono(i, toWarmup, startWarmupIteration, allocator).doOnSuccess(__ -> drain()))
-									.startWith(primary.doOnSuccess(__ -> drain()).then())
+									.startWith(primary.doOnSuccess(__ -> drain()).then().onErrorComplete())
```
Comment (pderop) [inline reactor-pool/src/main/java/reactor/pool/SimpleDequePool.java:449]:
```suggestion
									.startWith(primary.doOnSuccess(__ -> drain()).onErrorComplete().then())
```

although the patch seems correct, what do you think if we append `onErrorComplete` just after the `primary.doOnSuccess(...)`, it is just a matter of taste. I think it's the same, I just tend to prefer doing that in terms of clarity ?

@chemicL , do you agree with this patch ? and are you ok to move the onErrorComplete before `.then()`(just for readability reasons) ?

File: reactor-pool/src/test/java/reactor/pool/AllocationFailureHandlingTest.java [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+        StepVerifier.create(Flux.range(0, 10).concatMap(ignore -> pool.acquire().flatMap(PooledRef::invalidate)))
+            .verifyComplete();
+        assertThat(pool.metrics().idleSize()).isEqualTo(0);
+        assertThat(pool.config().allocationStrategy().estimatePermitCount()).isEqualTo(10);
+
+        // Now we have empty pool, so it should be warmed up again but allocator still not working
+        StepVerifier.create(pool.acquire()).verifyError();
+        assertThat(pool.metrics().idleSize()).isEqualTo(0);
+        assertThat(pool.config().allocationStrategy().estimatePermitCount()).isEqualTo(10); // Oops!
+    }
```
Comment (pderop) [inline reactor-pool/src/test/java/reactor/pool/AllocationFailureHandlingTest.java:65]: What do you think if we move the test in an existing test class, for example in `CommonPoolTest.java` ?
You could then rename your `test`method to `testIssue-174` for example ?

**Comment by k-tokarev** [comment]
> Here are few feedbacks :-)
> 
> @reactor/core-team, please take a look ? thanks

I committed all proposed changes. Can you review new version of Pull Request?

File: reactor-pool/src/test/java/reactor/pool/CommonPoolTest.java [hunk 4]
```diff
@@ -94,7 +94,7 @@ public <T, CONF extends PoolConfig<T>> AbstractPool<T> apply(PoolBuilder<T, CONF
 			return (AbstractPool<T>) builder.idleResourceReuseOrder(isLru).buildPool();
 		}
 	}
-	
+
```
Comment (pderop) [inline reactor-pool/src/test/java/reactor/pool/CommonPoolTest.java:97]: nit: can you remove the extra spaces in this line ?

Comment (k-tokarev) [reply reactor-pool/src/test/java/reactor/pool/CommonPoolTest.java:97]: > nit: can you remove the extra spaces in this line ?

It was automatically removed by my IDE. Do you mean restore extra spaces just to remove this diff? I will do it using GitHub editor (I don't want to disable this useful feature in my IDE).

Comment (pderop) [reply reactor-pool/src/test/java/reactor/pool/CommonPoolTest.java]: my bad, you are right, I was meaning restore the extra space, in order to keep the patch as minimal.

**Review by pderop — APPROVED** [review]
Looking good from my POV, but we need @reactor/core-team to also review, maybe they have more comments to do ?

File: reactor-pool/src/main/java/reactor/pool/SimpleDequePool.java [hunk 5]
```diff
@@ -451,7 +451,7 @@ else if (sig.isOnError()) {
 							final int mergeConcurrency = Math.min(poolConfig.allocationStrategy().warmupParallelism(), toWarmup + 1);
 							Flux.range(1, toWarmup)
 									.map(i -> warmupMono(i, toWarmup, startWarmupIteration, allocator).doOnSuccess(__ -> drain()))
-									.startWith(primary.doOnSuccess(__ -> drain()).then())
+									.startWith(primary.doOnSuccess(__ -> drain()).onErrorComplete().then())
```
Comment (chemicL) [inline reactor-pool/src/main/java/reactor/pool/SimpleDequePool.java:454]: As was discussed yesterday, this solves the issue for now but can be further improved to decouple unused permit return from connection acquisition. Right now it is implicit what happens and a reader of the code might scratch their head for some time to understand why this is used.

**Review by chemicL — APPROVED** [review]
Approving but hoping for a follow-up refactor to better structure the code and individual operators' responsibilities regarding connection acquisition, error handling, vs permit obtaining and returning to the pool.

**Comment by pderop** [comment]
@chemicL , thanks for the review, we'll try to  refactor all this in the next version.

@k-tokarev , thanks for the PR, which is merged.

**Pull request merged**